plugins/org.yakindu.sct.ui.editor/src/org/yakindu/sct/ui/editor/providers/StatechartValidationDecorationProvider.java

```diff
@@ -219,12 +238,27 @@ public void issuesChanged() {
 		}
 
 		@Override
-		public List<String> getSemanticURIs() {
-			List<String> result = new ArrayList<String>();
-			result.add(semanticID);
-			if (subdiagramSemanticID != null)
-				result.add(subdiagramSemanticID);
-			return result;
+		public Stream<String> getSemanticURIsStream() {
+
+			EObject element = getTargetEObject();
+			if (element == null) {
+				return Stream.of("-NO TARGET OBJECT-");
+			}
+
+			View view = getTargetView();
+			if (SemanticHints.STATE.equals(view.getType())) {
+				BooleanValueStyle style = GMFNotationUtil.getBooleanValueStyle(view,
+						DiagramPartitioningUtil.INLINE_STYLE);
+				if (style == null ? false : !style.isBooleanValue()) {
+					Stream<String> stream = StreamSupport.stream(() -> {
+						return Spliterators.spliteratorUnknownSize(element.eAllContents(), Spliterator.CONCURRENT);
+					}, Characteristics.CONCURRENT.ordinal(), true).map(e -> getSemanticURI(e));
```

Comment:
Traversing eAllContents() for all decorators each time a live validation is performed may be a performance problem. We should think of a more efficient solution